=pod

=head1 SYNOPSIS

switchseq -t get_switch [options]

=head1 OPTIONS

=over 8

=item B<--help> | B<-h>

Print the documentation.

=item B<** REQUIRED **>

=item B<--input> | B<-i> <file>

Input file. Columns have to be specified as follows:
gene_id (1)
transcript_id (2)
transcript expression levels in each sample (3-).
Eg, for an experiment with 3 controls and 3 KO, column names could be the following:

gene_id transcript_id C1 C2 C3 KO1 KO2 KO3

=item B<--cond1> | B<-c1> <interval>

Columns for condition 1. [example: --cond1 3-5]

=item B<--cond2> | B<-c2> <interval>

Columns for condition 2. [example: --cond2 6-8]

=item B<--species> | B<-s> <string>

Species (three letter code).

=item B<--ensembl_v> | B<-e> <integer>

Ensembl version.

=item B<** OPTIONAL **>

=item B<--out_dir> | B<-o> <dir>

html output directory. [default: ./html]

=item B<--data_dir> | B<-d> <dir>

Data directory containing information from Ensembl and APPRIS. [default: ./data]

=item B<--threshold_gexp> | B<-g> <float>

Gene expression threshold. [default: 0.01]

=item B<--filt> | B<-f> <file>

List of transcript ids that undergo differential usage across the studied conditions. [default: NA]

=back

=cut
