=pod

=head1 SYNOPSIS

lorem -t get_data [options]

=head1 OPTIONS

=over 8

=item B<--help> | B<-h>

Print the documentation.

=item B<** REQUIRED **>

=item B<--input> | B<-i> <file>

Input file. Columns have to be specified as follows:
gene_id (1)
transcript_id (2)
transcript FPKMs in each sample (3-).
For example, for an experiment with 3 controls and 3 KO, column names could be specified as follows:

gene_id transcript_id C1 C2 C3 KO1 KO2 KO3

=item B<--cond1> | B<-c1> <interval>

Columns for condition 1. [example: --cond1 3-5]

=item B<--cond2> | B<-c2> <interval>

Columns for condition 2. [example: --cond2 6-8]

=item B<--species> | B<-s> <hsa|mmu>

Species: human or mouse.

=item B<--ensembl_v> | B<-e> <integer>

Ensembl version.

=item B<** OPTIONAL **>

=item B<--output> | B<-o> <dir>

html output dir. [default: ./html]

=item B<--data_dir> | B<-d> <dir>

Temporary data directory. [default: ./data]

=item B<--threshold_gexp> | B<-g> <float>

Gene expression threshold. [default: 0.01]

=item B<--plot> | B<-p> <starplots|barplots|both>

Plots to create. [default: starplots]

=back

=cut