<!DOCTYPE html>
<html>
<head>
  	<title>Switch events</title>
	<link rel="stylesheet" type="text/css" href="./css/bootstrap.min.css">
	<link rel="stylesheet" type="text/css" href="./css/DT_bootstrap.css">
	<link rel="stylesheet" type="text/css" href="./css/TableTools.css">

	<script type="text/javascript" language="javascript" src="./js/jquery.js"></script>
	<script type="text/javascript" language="javascript" src="./js/jquery.dataTables.min.js"></script>
	<script type="text/javascript" language="javascript" src="./js/TableTools.min.js"></script>
	<script type="text/javascript" language="javascript" src="./js/bootstrap.min.js"></script>
	<script type="text/javascript" language="javascript" src="./js/DT_bootstrap.js"></script>
</head>

<body>
	<div class="container">
		<div class="accordion" id="accordion_options">
			<div class="accordion-group">
				<div class="accordion-heading">
				  <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion_options" href="#info">
				    <b>Execution information</b>
				  </a>
				</div>
			    <div id="info" class="accordion-body collapse">
				    <div class="accordion-inner">
				    	<p><b>Input:</b> ./data_test1/expdata.txt</p>
				    	<p><b>Data directory:</b> /Users/mar/Desktop/app/data_10</p>
				    	<p><b>Species:</b> hsa</p>
				    	<p><b>Ensembl version:</b> 66</p>
				    	<p><b>Condition 1:</b> 
				    		columns 3-6,
				    		4 samples
				    	</p>
				    	<p><b>Condition 2:</b> 
				    		columns 7-10,
				    		4 samples
				    	</p>
				 
				    </div>
			    </div>
		  	</div>

			 <div class="accordion-group">
			    <div class="accordion-heading">
			      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion_options" href="#filter">
			        <b>1. Filter</b>
			      </a>
			    </div>
			    <div id="filter" class="accordion-body collapse">
				    <div class="accordion-inner pagination-centered">

						<p><b>Functional classification of switch events<br>
							based on transcript biotype (C1, C2)</b></p>

						<table class="table" id="toc">
						<tbody>
							<tr>
								<td><b>All</></td>
								<td></td>
								<td><a href="index.html"></a></td>
							</tr>
							<tr>
								<td><b>Protein coding</b></td>
								<td></td>
								<td></td>
							</tr>
							<tr>
								<td>protein coding</td>
								<td>protein coding</td>
								<td><a href="pc_to_pc.html"></a></td>
							</tr>

							<tr>
								<td><b>Potential function loss</b></td>
								<td></td>
								<td></td>
							</tr>
							<tr>
								<td>protein coding</td>
								<td>nonsense mediated decay</td>
								<td><a href="pc_to_nmd.html"></a></td>
							</tr>
							<tr>
								<td>protein coding</td>
								<td>retained intron</td>
								<td><a href="pc_to_ri.html"></a></td>
							</tr>
							<tr>
								<td>protein coding</td>
								<td>processed transcript</td>
								<td><a href="pc_to_pt.html"></a></td>
							</tr>

							<tr>
								<td><b>Potential function gain</b></td>
								<td></td>
								<td></td>
							</tr>
							<tr>
								<td>nonsense mediated decay</td>
								<td>protein coding</td>
								<td><a href="nmd_to_pc.html"></a></td>
							</tr>
							<tr>
								<td>retained intron</td>
								<td>protein coding</td>
								<td><a href="ri_to_pc.html"></a></td>
							</tr>
								<tr><td>processed transcript</td>
								<td>protein coding</td>
								<td><a href="pt_to_pc.html"></a></td>
							</tr>
							<tr>
								<td><b>Other</b></td>
								<td></td>
								<td><a href="other.html"></a>
								</td>
							</tr>
						</tbody>
						</table>
					</div>
			    </div>
			</div>

		  	<div class="accordion-group">
				<div class="accordion-heading">
				  <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion_options" href="#format">
				    <b>2. Format</b>
				  </a>
				</div>
			    <div id="format" class="accordion-body collapse">
				    <div class="accordion-inner">
					<a href="javascript:void(0);" onclick="fnShowHide(0);"><input type="checkbox" checked /></a> <b>gId</b>: ensembl gene id<br><a href="javascript:void(0);" onclick="fnShowHide(1);"><input type="checkbox" checked /></a> <b>gName</b>: gene name<br><a href="javascript:void(0);" onclick="fnShowHide(2);"><input type="checkbox" checked /></a> <b>nOfT</b>: number of annotated transcripts<br><a href="javascript:void(0);" onclick="fnShowHide(3);"><input type="checkbox" checked /></a> <b>C1.tId</b>: major transcript - condition 1<br><a href="javascript:void(0);" onclick="fnShowHide(4);"><input type="checkbox" checked /></a> <b>C1.principal</b>: is the transcript classified as principal in APPRIS? - condition 1<br><a href="javascript:void(0);" onclick="fnShowHide(5);"><input type="checkbox" checked /></a> <b>C1.biotype</b>: major transcript biotype - condition 1<br><a href="javascript:void(0);" onclick="fnShowHide(6);"><input type="checkbox" checked /></a> <b>C1.tExp</b>: in how many samples is the transcript detected as major? - condition 1<br><a href="javascript:void(0);" onclick="fnShowHide(7);"><input type="checkbox" checked /></a> <b>C1.gExp</b>: in how many samples is the gene expressed? - condition 1<br><a href="javascript:void(0);" onclick="fnShowHide(8);"><input type="checkbox" checked /></a> <b>C1.breadth</b>: major transcript expression breadth - condition 1<br><a href="javascript:void(0);" onclick="fnShowHide(9);"><input type="checkbox" checked /></a> <b>C2.tId</b>: major transcript - condition 2<br><a href="javascript:void(0);" onclick="fnShowHide(10);"><input type="checkbox" checked /></a> <b>C2.principal</b>: is the transcript classified as principal in APPRIS? - condition 2<br><a href="javascript:void(0);" onclick="fnShowHide(11);"><input type="checkbox" checked /></a> <b>C2.biotype</b>: major transcript biotype - condition 2<br><a href="javascript:void(0);" onclick="fnShowHide(12);"><input type="checkbox" checked /></a> <b>C2.tExp</b>: in how many samples is the transcript detected as major? - condition 2<br><a href="javascript:void(0);" onclick="fnShowHide(13);"><input type="checkbox" checked /></a> <b>C2.gExp</b>: in how many samples is the gene expressed? - condition 2<br><a href="javascript:void(0);" onclick="fnShowHide(14);"><input type="checkbox" checked /></a> <b>C2.breadth</b>: major transcript expression breadth - condition 2<br><a href="javascript:void(0);" onclick="fnShowHide(15);"><input type="checkbox" checked /></a> <b>pIdentity</b>: percentage identity between the two coding sequences<br><a href="javascript:void(0);" onclick="fnShowHide(16);"><input type="checkbox" checked /></a> <b>pdbEntry</b>: is there any PDB entry available?<br><a href="javascript:void(0);" onclick="fnShowHide(17);"><input type="checkbox" checked /></a> <b>rank</b>: ranking to maximise expression breadth<br><a href="javascript:void(0);" onclick="fnShowHide(18);"><input type="checkbox" checked /></a> <b>boxplot</b>: link to boxplot<br><a href="javascript:void(0);" onclick="fnShowHide(19);"><input type="checkbox" checked /></a> <b>starplot</b>: link to starplot<br>
				    </div>
			    </div>
			</div>
		</div>
	</div>

	<div class="table" id="big_container">
		<table cellpadding="0" cellspacing="0" border="0" class="table table-striped table-bordered" id="main">
		<thead>
		  	<tr>
		  		<th>gId</th>
<th>gName</th>
<th>nOfT</th>
<th>C1.tId</th>
<th>C1.principal</th>
<th>C1.biotype</th>
<th>C1.tExp</th>
<th>C1.gExp</th>
<th>C1.breadth</th>
<th>C2.tId</th>
<th>C2.principal</th>
<th>C2.biotype</th>
<th>C2.tExp</th>
<th>C2.gExp</th>
<th>C2.breadth</th>
<th>pIdentity</th>
<th>pdbEntry</th>
<th>rank</th>
<th>boxplot</th>
<th>starplot</th>
 
			</tr>
		</thead>
		<tbody>
			<tr>
	<td><a href=http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000137409>ENSG00000137409</a></td>
	<td>MTCH1</td>
	<td>11</td>
	<td>ENST00000373616</td>
	<td><a href=http://appris.bioinfo.cnio.es/report.html?id=ENST00000373616&namespace=Ensembl_Transcript_Id>YES</a></td>
	<td>protein_coding</td>
	<td>4</td>
	<td>4</td>
	<td>100.00</td>
	<td>ENST00000373627</td>
	<td><a href=http://appris.bioinfo.cnio.es/report.html?id=ENST00000373627&namespace=Ensembl_Transcript_Id>NO</a></td>
	<td>protein_coding</td>
	<td>3</td>
	<td>4</td>
	<td>75.00</td>
<td><a href=./html_test1/prot_aln/ENSG00000137/ENSG00000137409.needle_mod.out>95.6</a></td>	<td>NO</td>
	<td>13.25</td>
<td><a href=./plots/boxplots/ENSG00000137/ENSG00000137409-rotated90.pdf>boxplot</a></td><td><a href=./plots/starplots/ENSG00000137/ENSG00000137409.pdf>starplot</a></td></tr>

		</tbody>
		</table>
	</div>
</body>
</html>